### PR TITLE
feat: Added staking pool class with balance

### PR DESCRIPTION
**Pull request by kj-crypto**

### CODE REVIEW

File: src/staking/staking_pool.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#ifndef ELECTRIC_CASH_STAKING_POOL_H
+#define ELECTRIC_CASH_STAKING_POOL_H
+#include "amount.h"
+
+class StakingPool {
```
Comment (mtwaro) [inline src/staking/staking_pool.h:5]: Naming convention of this codebase is to prefix every class name with a `C`, so this class should be called `CStakingPool`.

File: src/staking/staking_pool.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+private:
+    static StakingPool* srpInstance;
+    CAmount balance = 0;
+    StakingPool(){};
```
Comment (mtwaro) [inline src/staking/staking_pool.h:10]: I think we're going to need a constructor with the balance as a parameter in near future.

File: src/staking/staking_pool.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+public:
+    StakingPool(StakingPool& copy) = delete;
```
Comment (mtwaro) [inline src/staking/staking_pool.h:13]: Singleton classes also shouldn't be be assignable
`void operator=(const StakingPool&) = delete;`

File: src/consensus/block_rewards.cpp [hunk 4]
```diff
@@ -11,4 +12,10 @@ CAmount GetBlockRewardForHeight(uint32_t height)
          }
      }
      return 0;
- }
\ No newline at end of file
+ }
+
+CAmount GetStakingRewardForHeight(uint32_t height)
+{
+    return (CAmount) floor(FRACTION_OF_STAKING_REWARD * GetBlockRewardForHeight(height));
```
Comment (mtwaro) [inline src/consensus/block_rewards.cpp:19]: Better to use `static_cast`

File: src/miner.cpp [hunk 5]
```diff
@@ -151,6 +155,7 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = scriptPubKeyIn;
     coinbaseTx.vout[0].nValue = nFees + GetBlockSubsidy(nHeight);
+    StakingPool::getInstance()->increaseBalance(GetStakingRewardForHeight(nHeight));
```
Comment (mtwaro) [inline src/miner.cpp:158]: Staking pool should be modified only if a new, already mined block appears. We shouldn't increase it when assembling a new block.

File: src/staking/staking_pool.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#include "staking_pool.h"
```
Comment (mtwaro) [inline src/staking/staking_pool.cpp:1]: Should work with `<>`

File: src/staking/staking_pool.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    static StakingPool* getInstance();
+    void increaseBalance(CAmount amount);
+    void decreaseBalance(CAmount amount);
```
Comment (mtwaro) [inline src/staking/staking_pool.h:16]: I think also `setBalance` will be needed.

**Review by mtwaro — CHANGES_REQUESTED** [review]
Need to add staking pool manipulation logic to `validation.cpp`. to block validation. Also, reorgs have to be considered.

File: src/staking/staking_pool.h (hunk 1 above)
Comment (kj-crypto) [reply src/staking/staking_pool.h]: done

File: src/staking/staking_pool.h (hunk 3 above)
Comment (kj-crypto) [reply src/staking/staking_pool.h]: done

File: src/consensus/block_rewards.cpp (hunk 4 above)
Comment (kj-crypto) [reply src/consensus/block_rewards.cpp]: done

File: src/staking/staking_pool.cpp (hunk 6 above)
Comment (kj-crypto) [reply src/staking/staking_pool.cpp]: done

File: src/staking/staking_pool.h (hunk 7 above)
Comment (kj-crypto) [reply src/staking/staking_pool.h:21]: done

File: src/init.cpp [hunk 8]
```diff
@@ -611,6 +609,12 @@ static void BlockNotifyCallback(bool initialSync, const CBlockIndex *pBlockIndex
 }
 #endif
 
+static void StakingPoolCallback(bool initialSync, const CBlockIndex *pBlockIndex) {
+    if (initialSync || !pBlockIndex)
+        return;
+    CStakingPool::getInstance()->increaseBalanceForNewBlock(pBlockIndex->nHeight);
+}
```
Comment (mtwaro) [inline src/init.cpp:616]: That's not a good place for this. I think the best place to increment staking pool balance would be at `CChainState::ConnectBlock`. Moreover, CChainstate would be the best handler for staking pool as a whole

File: src/staking/staking_pool.h (hunk 2 above)
Comment (kj-crypto) [reply src/staking/staking_pool.h]: Added

File: src/init.cpp (hunk 8 above)
Comment (kj-crypto) [reply src/init.cpp]: I moved increase and decrease staking balance feature into `ConnectBlock` and `DisconnectBlock` methods. I added also functional test in `feature_staking_pool.py` to check correctness of the implementation.

**Review by mtwaro — COMMENTED** [review]
Fine by me, however tests have to pass before we merge this.

**Review by mtwaro — APPROVED** [review]

**Pull request merged**